### DISCUSSION TITLE
NextJS application that uses pages API routes

### DISCUSSION THREAD

**Discussion by ebrearley** [coollabsio/coolify#2718] [Q&A]
I'm currently in the process of migrating a NextJS application to Coolify and have hit a bunch of problems with [API routes](https://nextjs.org/docs/pages/building-your-application/routing/api-routes), in particular those which rely on the published path of the application like [Next Auth](https://next-auth.js.org/getting-started/example).

From what I can see, the Nixpacks build is running the `build` and `start` scripts from `package.json`, exactly as I would if the application was directly exposed to the internet, rather than behind a reverse proxy. This means a package like NextAuth will respond with a redirect to `http://localhost:3000` (the application in the container), rather than the domain configured in the Coolify resource.

For API routes to work correctly in Coolify, it seems the NextJS application needs to started up in a way that is aware of it's hostname (maybe a [custom server](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server) is required to pull this off - which does have some tradeoffs like disabling automatic static optimization of routes). Or is there something else I can do with Traefik, like some path rewriting? Although rewrites might still lead to unexpected behaviour in API routes if the API route does anything with the hostname of the incoming request. 

Has anyone else run into this problem? Or are your API routes working as expected?

**Comment by ebrearley** [comment]
I worked around this by defining the base path of the origin domain as an environment variable and respond with that. Outside of this, my API routes work as expected.

**Reply by 90PabloRomero** [reply]
sorry, could you explain exactly how you done this?

**Reply by mirsella** [reply]
hey, would be really helpful if you could give us more information ! have a good day
